File: documentation/AllMetaheuristic/Evolutionary-based/memetic_algorithm.py
import numpy as np

class MemeticAlgorithm:
    def __init__(self, objective_function, dim=2, bounds=None,
                 pop_size=20, max_iterations=100,
                 mutation_rate=0.1, crossover_rate=0.7):
        """
        Initialize the Memetic Algorithm optimizer.

        Parameters:
        - objective_function: Function to be minimized.
        - dim: Number of decision variables.
        - bounds: List of (lower, upper) tuples for each dimension.
        - pop_size: Size of the population.
        - max_iterations: Maximum number of iterations.
        - mutation_rate: Probability of mutation.
        - crossover_rate: Proportion of population used for crossover.
        """
        self.obj_func = objective_function
        self.dim = dim
        self.bounds = bounds if bounds else [(-5, 5)] * dim
        self.pop_size = pop_size
        self.max_iterations = max_iterations
        self.mutation_rate = mutation_rate
        self.crossover_rate = crossover_rate

        self.lower_bounds = np.array([b[0] for b in self.bounds])
        self.upper_bounds = np.array([b[1] for b in self.bounds])

        self.population = np.random.uniform(self.lower_bounds, self.upper_bounds, (self.pop_size, self.dim))
        self.fitness = np.apply_along_axis(self.obj_func, 1, self.population)

        self.best_solution = None
        self.best_fitness = float("inf")
        self.history = []

    def optimize(self):
        for iteration in range(self.max_iterations):
            # Local search (Hill Climbing)
            local_search_count = max(1, int(0.1 * self.pop_size))
            for _ in range(local_search_count):
                idx = np.random.randint(self.pop_size)
                new_ind = self.hill_climbing(self.population[idx])
                new_fit = self.obj_func(new_ind)
                if new_fit < self.fitness[idx]:
                    self.population[idx] = new_ind
                    self.fitness[idx] = new_fit

            # Crossover
            sorted_indices = np.argsort(self.fitness)
            num_parents = max(2, int(self.crossover_rate * self.pop_size))
            parent_indices = sorted_indices[:num_parents]
            children = []

            for i in range(0, num_parents - 1, 2):
                p1 = self.population[parent_indices[i]]
                p2 = self.population[parent_indices[i + 1]]
                offspring = self.crossover(p1, p2)
                children.extend(offspring)

            children = np.array(children)

            # Mutation
            for i in range(len(children)):
                if np.random.rand() < self.mutation_rate:
                    children[i] = self.mutation(children[i])

            # Replace worst individuals
            replace_count = min(len(children), self.pop_size)
            worst_indices = np.argsort(-self.fitness)[:replace_count]
            self.population[worst_indices] = children[:replace_count]

            for i in range(replace_count):
                self.fitness[worst_indices[i]] = self.obj_func(self.population[worst_indices[i]])

            # Update best solution
            current_best_idx = np.argmin(self.fitness)
            if self.fitness[current_best_idx] < self.best_fitness:
                self.best_fitness = self.fitness[current_best_idx]
                self.best_solution = self.population[current_best_idx].copy()

            self.history.append((iteration, self.best_solution.copy()))
            print(f"Iteration {iteration + 1}: Best Fitness = {self.best_fitness}")

        return self.best_solution, self.best_fitness, self.history

    def hill_climbing(self, initial_solution):
        current_solution = initial_solution.copy()
        current_fitness = self.obj_func(current_solution)

        for _ in range(100):
            perturbation = 0.1 * np.random.randn(self.dim)
            new_solution = np.clip(current_solution + perturbation, self.lower_bounds, self.upper_bounds)
            new_fitness = self.obj_func(new_solution)

            if new_fitness < current_fitness:
                current_solution = new_solution
                current_fitness = new_fitness

        return current_solution

    def crossover(self, parent1, parent2):
        point = np.random.randint(1, self.dim)
        offspring1 = np.concatenate([parent1[:point], parent2[point:]])
        offspring2 = np.concatenate([parent2[:point], parent1[point:]])
        return [offspring1, offspring2]

    def mutation(self, solution):
        mutated = solution + 0.1 * np.random.randn(self.dim)
        return np.clip(mutated, self.lower_bounds, self.upper_bounds)

File: documentation/AllMetaheuristic/Evolutionary-based/memetic algorithm/memetic algorithm/memetic_algorithm.m
function [best_solution, best_fitness] = memetic_algorithm(obj_func, dim, lower_bounds, upper_bounds, max_iter, pop_size, mutation_rate, crossover_rate)
    % Initialize population
    population = lower_bounds + (upper_bounds - lower_bounds) .* rand(pop_size, dim);
    
    % Evaluate initial population
    fitness = zeros(pop_size, 1);
    for i = 1:pop_size
        fitness(i) = obj_func(population(i, :));
    end
    
    % Main loop
    for iter = 1:max_iter
        % Perform local search (hill climbing) on a portion of the population
        local_search_pop_size = round(0.1 * pop_size); % adjust this fraction as needed
        for i = 1:local_search_pop_size
            % Perform hill climbing starting from a random individual
            individual = population(randi(pop_size), :);
            new_individual = hill_climbing(obj_func, individual, lower_bounds, upper_bounds);
            
            % Replace the individual if the new one is better
            new_fitness = obj_func(new_individual);
            if new_fitness < fitness(i)
                population(i, :) = new_individual;
                fitness(i) = new_fitness;
            end
        end
        
        % Select parents for crossover
        [~, sorted_indices] = sort(fitness);
        parent_indices = sorted_indices(1:round(crossover_rate * pop_size));
        
        % Perform crossover
        children = zeros(length(parent_indices), dim);
        for i = 1:2:length(parent_indices)
            parent1 = population(parent_indices(i), :);
            parent2 = population(parent_indices(i+1), :);
            children(i:i+1, :) = crossover(parent1, parent2);
        end
        
        % Perform mutation
        for i = 1:length(parent_indices)
            if rand < mutation_rate
                children(i, :) = mutation(children(i, :), lower_bounds, upper_bounds);
            end
        end
        
        % Replace worst individuals with offspring
        [~, worst_indices] = max(fitness);
        num_worst = length(worst_indices);
        population(worst_indices, :) = children(1:num_worst, :);
        
        % Update fitness values
        for i = 1:num_worst
            fitness(worst_indices(i)) = obj_func(population(worst_indices(i), :));
        end
    end
    
    % Find the best solution
    [best_fitness, best_index] = min(fitness);
    best_solution = population(best_index, :);
end

function new_solution = hill_climbing(obj_func, initial_solution, lower_bounds, upper_bounds)
    current_solution = initial_solution;
    current_fitness = obj_func(current_solution);
    
    % Perform hill climbing
    max_iter = 100; % adjust as needed
    for iter = 1:max_iter
        % Generate a new solution by perturbing the current one
        new_solution = current_solution + 0.1 * randn(size(current_solution));
        % Ensure the new solution is within bounds
        new_solution = max(lower_bounds, min(upper_bounds, new_solution));
        
        % If the new solution is better, replace the current one
        new_fitness = obj_func(new_solution);
        if new_fitness < current_fitness
            current_solution = new_solution;
            current_fitness = new_fitness;
        end
    end
    
    new_solution = current_solution;
end

function offspring = crossover(parent1, parent2)
    % Single point crossover
    crossover_point = randi([1, length(parent1)]);
    offspring1 = [parent1(1:crossover_point), parent2(crossover_point+1:end)];
    offspring2 = [parent2(1:crossover_point), parent1(crossover_point+1:end)];
    offspring = [offspring1; offspring2];
end

function mutated_solution = mutation(solution, lower_bounds, upper_bounds)
    % Random mutation
    mutation_range = 0.1; % adjust as needed
    mutated_solution = solution + mutation_range * randn(size(solution));
    % Ensure the mutated solution is within bounds
    mutated_solution = max(lower_bounds, min(upper_bounds, mutated_solution));
end
